spec/lib/rmt/downloader_spec.rb

```diff
@@ -0,0 +1,137 @@
+          .with(headers: headers)
+          .to_return(status: 200, body: 'test', headers: {})
+      end
+
+      context 'and hash function is unknown' do
+        it 'raises an exception' do
+          expect do
+            downloader.download('/repomd.xml', 'CHUNKYBACON42', '0xDEADBEEF')
+          end.to raise_error(RMT::Downloader::Exception, 'Unknown hash function CHUNKYBACON42')
+        end
+      end
+
+      context 'and checksum is wrong' do
+        it 'raises an exception' do
+          expect do
+            downloader.download('/repomd.xml', 'SHA256', '0xDEADBEEF')
+          end.to raise_error(RMT::Downloader::Exception, 'Checksum doesn\'t match')
+        end
+      end
+
+      context 'and checksum is correct' do
+        subject { downloader.download('/repomd.xml', checksum_type, checksum) }
+
+        it('has correct content') { expect(File.read(subject)).to eq(content) }
+      end
+    end
+  end
+
+  describe '#download_multi' do
+    let(:files) { %w(package1 package2 package3) }
+    let(:checksum_type) { 'SHA256' }
+    let(:queue) do
+      queue = []
+      files.each do |file|
+        queue << RMT::Rpm::FileEntry.new(
+          file,
+          checksum_type,
+          Digest.const_get(checksum_type).hexdigest(file),
+          :rpm
+        )
+      end
+      queue
+    end
+
+    before do
+      files.each do |file|
+        stub_request(:get, "http://example.com/#{file}").with(headers: headers)
+          .to_return(status: 200, body: file, headers: {})
+      end
+      downloader.download_multi(queue)
+    end
+
+    it 'requested all files' do
+      files.each do |file|
+        expect(WebMock).to(
+          have_requested(:get, "http://example.com/#{file}").with(headers: headers)
+        )
+      end
+    end
+
+    it 'saved all files' do
+      files.each do |file|
+        content = File.read(File.join(dir, file))
+        expect(content).to eq(file)
+      end
+    end
+  end
+
+  describe '#download_multi handles exceptions properly' do
+    let(:files) { %w(package1 package2 package3) }
+    let(:checksum_type) { 'SHA256' }
+    let(:queue) do
+      queue = []
+      files.each do |file|
+        queue << RMT::Rpm::FileEntry.new(
+          file,
+          checksum_type,
+          Digest.const_get(checksum_type).hexdigest(file),
+          :rpm
+        )
+      end
+      queue
+    end
+
+    before do
+      files.each do |file|
+        stub_request(:get, "http://example.com/#{file}").with(headers: headers)
+          .to_return(status: 404, body: file, headers: {})
+      end
+      downloader.download_multi(queue)
+    end
+
+    it 'requested all files' do
+      files.each do |file|
+        expect(WebMock).to(
+          have_requested(:get, "http://example.com/#{file}").with(headers: headers)
+        )
+      end
+    end
+
+    it 'saved all files' do
+      files.each do |file|
+        expect(File.exist?(File.join(dir, file))).to eq(false)
```

Comment:
Whoops, example description is incorrect. It is supposed to check that it doesn't write files if there were exceptions raised.